File: examples/ospf-kotlin-example/src/main/fuookami/ospf/kotlin/example/framework_demo/demo4/flight_task_context/model/Aircraft.kt
package fuookami.ospf.kotlin.example.framework_demo.demo4.flight_task_context.model

import kotlin.time.*
import kotlinx.datetime.*
import fuookami.ospf.kotlin.utils.math.*
import fuookami.ospf.kotlin.example.framework_demo.demo4.infrastructure.*
import fuookami.ospf.kotlin.framework.gantt_scheduling.domain.task.model.*

enum class AircraftCategory {
    Passenger,
    Cargo
}

sealed class AircraftCapacity {
    class Passenger(
        private val capacity: Map<PassengerClass, UInt64>
    ) : AircraftCapacity() {
        val total: UInt64 = UInt64(capacity.values.sumOf { it.toInt() }.toULong())

        operator fun get(cls: PassengerClass): UInt64 {
            return capacity[cls] ?: UInt64.zero
        }

        fun enabled(payload: Map<PassengerClass, UInt64>): Boolean {
            return payload.asSequence().all { this[it.key] >= it.value }
        }

        override val category: AircraftCategory get() = AircraftCategory.Passenger
    }

    class Cargo(
        val capacity: Flt64
    ) : AircraftCapacity() {
        override val category: AircraftCategory get() = AircraftCategory.Cargo

        fun enabled(payload: Flt64): Boolean {
            return capacity geq payload
        }
    }

    abstract val category: AircraftCategory
}

data class AircraftType(
    val code: AircraftTypeCode
) {
    companion object {
        private val pool = HashMap<AircraftTypeCode, AircraftType>()
        val values by pool::values

        operator fun invoke(code: AircraftTypeCode): AircraftType {
            return pool.getOrPut(code){ AircraftType(code) }
        }
    }

    override fun hashCode(): Int {
        return code.hashCode()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as AircraftType

        return code == other.code
    }
}

operator fun Map<Route, Duration>.get(dep: Airport, arr: Airport): Duration? {
    return this[Route(dep, arr)]
}

data class AircraftMinorType(
    val type: AircraftType,
    val code: AircraftMinorTypeCode,
    val costPerHour: Flt64,
    val routeFlyTime: Map<Route, Duration>,
    val connectionTime: Map<Airport, Duration>,
    val maxFlyTime: Duration? = null
) {
    val maxRouteFlyTime: Duration by lazy { routeFlyTime.asSequence().maxOf { it.value } }
    val maxConnectionTime: Duration by lazy { connectionTime.asSequence().maxOf { it.value } }

    companion object {
        private val pool = HashMap<AircraftMinorTypeCode, AircraftMinorType>()
        val values by pool::values

        operator fun invoke(code: AircraftMinorTypeCode): AircraftMinorType? {
            return pool[code]
        }
    }

    override fun hashCode(): Int {
        return type.hashCode().inv() or code.hashCode()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as AircraftMinorType

        if (type != other.type) return false
        if (code != other.code) return false

        return true
    }

    override fun toString() = "$code"
}

data class Aircraft(
    val regNo: AircraftRegisterNumber,
    val minorType: AircraftMinorType,
    val capacity: AircraftCapacity
) : Executor(regNo.no, regNo.no) {
    internal lateinit var _usability: AircraftUsability
    val usability get() = _usability

    val type by minorType::type
    val costPerHour by minorType::costPerHour
    val routeFlyTime by minorType::routeFlyTime
    val maxFlyTime by minorType::maxFlyTime
    val maxRouteFlyTime by minorType::maxRouteFlyTime
    val connectionTime by minorType::connectionTime
    val maxConnectionTime by minorType::maxConnectionTime

    companion object {
        private val pool = HashMap<AircraftRegisterNumber, Aircraft>()
        val values by pool::values

        operator fun invoke(regNo: AircraftRegisterNumber): Aircraft? {
            return pool[regNo]
        }
    }

    init {
        pool[regNo] = this
    }

    fun capacity(cls: PassengerClass): UInt64 {
        return when (val capacity = this.capacity) {
            is AircraftCapacity.Passenger -> { capacity[cls] }
            else -> { UInt64.zero }
        }
    }

    override fun toString() = "$regNo"
}

class AircraftUsability(
    lastTask: FlightTask?,
    val location: Airport,
    enabledTime: Instant,
    val flightCyclePeriods: List<FlightCyclePeriod> = emptyList()
) : ExecutorInitialUsability<FlightTask, Aircraft, FlightTaskAssignment>(lastTask, enabledTime) {
    fun overFlightHourTimes(time: Instant, flightHour: FlightHour): UInt64 {
        return UInt64(flightCyclePeriods.count {
            time < it.expirationTime && !it.enabled(flightHour)
        })
    }

    fun overFlightCycleTimes(time: Instant, flightCycle: FlightCycle): UInt64 {
        return UInt64(flightCyclePeriods.count {
            time < it.expirationTime && !it.enabled(flightCycle)
        })
    }

    fun overFlightHour(time: Instant, flightHour: FlightHour): FlightHour {
        var ret = FlightHour.zero
        for (period in flightCyclePeriods) {
            if (time >= period.expirationTime) {
                continue
            }
            ret += period.overFlightHour(flightHour)
        }
        return ret
    }

    fun overFlightCycle(time: Instant, flightCycle: FlightCycle): FlightCycle {
        var ret = FlightCycle.zero
        for (period in flightCyclePeriods) {
            if (time >= period.expirationTime) {
                continue
            }
            ret += period.overFlightCycle(flightCycle)
        }
        return ret
    }
}

File: ospf-rust-example/src/framework/demo4/domain/task/model/aircraft.rs
//! 飞机模型模块 / Aircraft model module

use time::{Duration, OffsetDateTime};
use ospf_rust_framework_gantt_scheduling::domain::task::ExecutorTrait;
use crate::framework::demo4::infrastructure::{AircraftRegisterNumber, PassengerClass};
use super::aircraft_type::AircraftMinorType;
use super::airport::Airport;
use super::flight_cycle::FlightCyclePeriod;

/// 飞机容量 / Aircraft capacity
#[derive(Debug, Clone)]
pub enum AircraftCapacity {
    /// 客机容量 / Passenger aircraft capacity
    Passenger {
        /// 各舱位等级容量列表 / Capacity per passenger class
        capacities: Vec<(PassengerClass, u64)>,
    },
    /// 货机容量 / Cargo aircraft capacity
    Cargo {
        /// 最大载重 / Maximum cargo weight
        max_weight: f64,
    },
}

/// 飞机 / Aircraft
#[derive(Debug, Clone)]
pub struct Aircraft {
    /// 注册号 / Registration number
    pub reg_no: AircraftRegisterNumber,
    /// 飞机子类型 / Aircraft minor type
    pub minor_type: AircraftMinorType,
    /// 容量 / Capacity
    pub capacity: AircraftCapacity,
}

impl Aircraft {
    /// 获取飞机类型代码 / Get aircraft type code
    pub fn type_code(&self) -> &super::aircraft_type::AircraftType {
        &self.minor_type.aircraft_type
    }

    /// 获取每小时成本 / Get cost per flight hour
    pub fn cost_per_hour(&self) -> f64 {
        self.minor_type.cost_per_hour
    }
}

impl PartialEq for Aircraft {
    fn eq(&self, other: &Self) -> bool {
        self.reg_no == other.reg_no
    }
}
impl Eq for Aircraft {}

impl std::hash::Hash for Aircraft {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.reg_no.hash(state);
    }
}

/// 实现 ExecutorTrait for Aircraft
impl ExecutorTrait for Aircraft {
    type Id = String;

    fn id(&self) -> &Self::Id {
        &self.reg_no.0
    }

    fn name(&self) -> &str {
        &self.reg_no.0
    }
}

/// 飞机可用性 / Aircraft usability
#[derive(Debug, Clone)]
pub struct AircraftUsability {
    /// 当前所在机场 / Current location airport
    pub location: Airport,
    /// 可用时间 / Enabled time
    pub enabled_time: OffsetDateTime,
    /// 飞行循环周期列表 / Flight cycle periods
    pub flight_cycle_periods: Vec<FlightCyclePeriod>,
}
